File: dynamic_programming/buy_and_sell_stock/121_best_time_to_buy_and_sell_stock.cc
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

// 2024/01/05: 自己写出来贪心算法, 但是是逆序. 维护未来最高价格.

// 只可以交易一次
int MaxProfit(const std::vector<int>& prices) {
  int history_min_price = std::numeric_limits<int>::max();
  int max_profit = 0;
  for (int i = 0; i < static_cast<int>(prices.size()); ++i) {
    max_profit = std::max(max_profit, prices[i] - history_min_price);
    history_min_price = std::min(history_min_price, prices[i]);
  }
  return max_profit;
}

int main(int argc, char* argv[]) {
  std::vector<int> prices = {7, 1, 5, 3, 6, 4};
  std::cout << MaxProfit(prices) << std::endl;
  return 0;
}

File: dynamic_programming/buy_and_sell_stock/121_best_time_to_buy_and_sell_stock.py
from typing import List


# 2026/06/04: 自己写出答案.
def max_profit(prices: List[int]) -> int:
    dp = [0] * len(prices)  # dp[i] 表示以 i 结尾的最大利润.

    min_price = prices[0]
    for i in range(1, len(prices)):
        dp[i] = max(dp[i - 1], prices[i] - min_price)
        min_price = min(min_price, prices[i])

    return dp[-1]


prices1 = [7, 1, 5, 3, 6, 4]
prices2 = [7, 6, 4, 3, 1]
prices3 = [1, 2]
print(max_profit(prices3))
